src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,344 @@
+        boost::array<uint256, NumOutputs>& commitments,
+        uint64_t vpub_old,
+        uint64_t vpub_new,
+        const uint256& rt
+    ) {
+        if (!pk) {
+            throw std::runtime_error("JoinSplit proving key not loaded");
+        }
+
+        // Compute nullifiers of inputs
+        for (size_t i = 0; i < NumInputs; i++) {
+            nullifiers[i] = inputs[i].nullifier();
+        }
+
+        // Sample randomSeed
+        randomSeed = random_uint256();
+
+        // Compute h_sig
+        uint256 h_sig = this->h_sig(randomSeed, nullifiers, pubKeyHash);
+
+        // Sample phi
+        uint256 phi = random_uint256();
+
+        // Compute notes for outputs
+        for (size_t i = 0; i < NumOutputs; i++) {
+            // Sample r
+            uint256 r = random_uint256();
+
+            output_notes[i] = outputs[i].note(phi, r, i, h_sig);
+        }
+
+        // Compute the output commitments
+        for (size_t i = 0; i < NumOutputs; i++) {
+            commitments[i] = output_notes[i].cm();
+        }
+
+        // Encrypt the ciphertexts containing the note
+        // plaintexts to the recipients of the value.
+        {
+            ZCNoteEncryption encryptor(h_sig);
+
+            for (size_t i = 0; i < NumOutputs; i++) {
+                // TODO: expose memo in the public interface
+                boost::array<unsigned char, ZCASH_MEMO_SIZE> memo;
+                memo[0] = 0xF6; // invalid UTF8 as per spec
+
+                NotePlaintext pt(output_notes[i], memo);
+
+                ciphertexts[i] = pt.encrypt(encryptor, outputs[i].addr.pk_enc);
+            }
+
+            ephemeralKey = encryptor.get_epk();
+        }
+
+        // Authenticate h_sig with each of the input
+        // spending keys, producing hmacs which protect
+        // against malleability.
+        for (size_t i = 0; i < NumInputs; i++) {
+            hmacs[i] = PRF_pk(inputs[i].key, i, h_sig);
+        }
+
+        std::vector<FieldT> primary_input;
+        std::vector<FieldT> aux_input;
+
+        {
+            protoboard<FieldT> pb;
+            {
+                joinsplit_gadget<FieldT, NumInputs, NumOutputs> g(pb);
+                g.generate_r1cs_constraints();
+                g.generate_r1cs_witness(
+                    phi,
+                    rt,
+                    h_sig,
+                    inputs,
+                    output_notes,
+                    vpub_old,
+                    vpub_new
+                );
+            }
+
+            if (!pb.is_satisfied()) {
+                throw std::invalid_argument("Constraint system not satisfied by inputs");
+            }
+
+            primary_input = pb.primary_input();
+            aux_input = pb.auxiliary_input();
+        }
+
+        auto proof = r1cs_ppzksnark_prover<ppzksnark_ppT>(
+            *pk,
+            primary_input,
+            aux_input
+        );
+
+        std::stringstream ss;
+        ss << proof;
+
+        return ss.str();
```

Comment:
There's no expected length of the proof because it's encoded in ascii decimal.

However, [in my future PR](https://github.com/ebfull/zcash/commit/8ff0dd4c2d3ebe25d3ddf2ebfa21b07454b44462), I have implemented fixed-sized proofs and wrote such assertions, because I enabled binary serialization.